File: ProblemSolve/938.cpp
/********************
 * Range Sum of BST *
 ********************/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int rangeSumBST(TreeNode* root, int low, int high, int sum = 0) {
        if(root == NULL){
            return sum;
        }
                
        if(root -> val < low){
            //BST이기 때문에 left subtree의 모든 node 값은 범위 밖의 값이다
            sum = rangeSumBST(root -> right, low, high, sum);
        }        
        else if(root -> val > high){
            //BST이기 때문에 right subtree의 모든 node 값은 범위 밖의 값이다        
            sum = rangeSumBST(root -> left, low, high, sum);
        }
        else{
            sum += root -> val;
            sum = rangeSumBST(root -> right, low, high, sum);
            sum = rangeSumBST(root -> left, low, high, sum);
        }
        
        return sum;
    }
};

File: ProblemSolve/938.cs
/********************
 * Range Sum of BST *
 ********************/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    int sum = 0;
    
    public int RangeSumBST(TreeNode root, int low, int high) {
        if(root == null){
            return sum;
        }
        
        //BST이기 때문에 현재 값이 범위 밖에 있으면 일부 범위를 탐색하지 않아도 됨.
        if(root.val < low){
            sum = RangeSumBST(root.right, low, high);
        }
        else if(root.val > high){
            sum = RangeSumBST(root.left, low, high);
        }
        else{
            sum += root.val;
            sum = RangeSumBST(root.left, low, high);
            sum = RangeSumBST(root.right, low, high);
        }
        
        return sum;
    }
}